File: frontend/foyle/src/web/converters.ts
// Conversion routines between the protocol buffer format and VSCode's representation of a notebook
//
// See ../vscode_apis.md for an exlanation. It is very helpful for understanding this folder.

import * as vscode from 'vscode';
import * as docpb from "../gen/foyle/v1alpha1/doc_pb";
import * as metadata from "./metadata";
import * as constants from "./constants";

// cellToCellData converts a NotebookCell to a NotebookCellData.
// NotebookCell is an interface used by the editor.
// NotebookCellData is a concrete class.
export function cellToCellData(cell: vscode.NotebookCell): vscode.NotebookCellData {
  let data = new vscode.NotebookCellData(cell.kind, cell.document.getText(), cell.document.languageId);

  data.metadata = cell.metadata;
  data.outputs = [];
  for (let o of cell.outputs) {
    data.outputs.push(o);
  }
  return data;
}

// cellDataToBlock converts an instance of NotebookCellData to the Block proto.
export function cellDataToBlock(cell: vscode.NotebookCellData): docpb.Block {
  let block = new docpb.Block();
  block.contents = cell.value;
  block.language = cell.languageId;
  if (cell.kind === vscode.NotebookCellKind.Code) {
    block.kind = docpb.BlockKind.CODE;
  } else {
    block.kind = docpb.BlockKind.MARKUP;
  }

  if (cell.metadata !== undefined) {
    metadata.setBlockFromMeta(block, cell.metadata);
  }
  
  let cellOutputs: vscode.NotebookCellOutput[] = [];
  if (cell.outputs !== undefined) {
    cellOutputs = cell.outputs;
  }
  for (const output of cellOutputs) {
    let outBlock = new docpb.BlockOutput();
    outBlock.items = [];
    for (const item of output.items) {
      let outItem = new docpb.BlockOutputItem();
      outItem.textData = new TextDecoder().decode(item.data);
      outItem.mime = item.mime;
      outBlock.items.push(outItem);
    }
    block.outputs.push(outBlock);
  }

  return block;
}

// blockToCellData converts a Block proto to a NotebookCellData.
export function blockToCellData(block: docpb.Block): vscode.NotebookCellData {
  let kind = vscode.NotebookCellKind.Markup;
  if (block.kind === docpb.BlockKind.CODE) { 
    kind = vscode.NotebookCellKind.Code;
  }
  let language = block.language;
  if (language === "" && kind === vscode.NotebookCellKind.Code) {
    // Default to the bash language
    language = constants.bashLang;

    if (language !== constants.bashLang) {
      console.log(`Unsupported language: ${language}`);
    }
  }

  let newCell = new vscode.NotebookCellData(
    kind,
    block.contents,
    language             
  );

  newCell.metadata = metadata.getCellMetadata(block);
  newCell.outputs = [];
  for (let output of block.outputs) {
    let items: vscode.NotebookCellOutputItem[] = [];
    for (let item of output.items) {
      if (item.textData !== "") {
        const text = new TextEncoder().encode(item.textData);            
        items.push(new vscode.NotebookCellOutputItem(text, item.mime));          
      } else {
        console.log("Unknown output type");
      }
    }
    newCell.outputs.push(new vscode.NotebookCellOutput(items));
  }
  return newCell;
}

File: app/pkg/docs/converters.go
package docs

import (
	"math"
	"strings"

	"github.com/jlewi/foyle/app/pkg/runme/converters"
	parserv1 "github.com/stateful/runme/v3/pkg/api/gen/proto/go/runme/parser/v1"
	"github.com/stateful/runme/v3/pkg/document/identity"

	"github.com/jlewi/foyle/protos/go/foyle/v1alpha1"
	"github.com/stateful/runme/v3/pkg/document/editor"
)

const (
	codeTruncationMessage = "<...code was truncated...>"
	truncationMessage     = "<...stdout was truncated...>"
)

// BlockToMarkdown converts a block to markdown
// maxLength is a maximum length for the generated markdown. This is a soft limit and may be exceeded slightly
// because we don't account for some characters like the outputLength and the truncation message
// A value <=0 means no limit.
func BlockToMarkdown(block *v1alpha1.Block, maxLength int) string {
	sb := strings.Builder{}
	writeBlockMarkdown(&sb, block, maxLength)
	return sb.String()
}

func writeBlockMarkdown(sb *strings.Builder, block *v1alpha1.Block, maxLength int) {

	maxInputLength := -1
	maxOutputLength := -1

	if maxLength > 0 {
		// Allocate 50% of the max length for input and output
		// This is crude. Arguably we could be dynamic e.g. if the output is < .5 maxLength we should allocate
		// the unused capacity for inputs. But for simplicity we don't do that. We do allocate unused input capacity
		// to the output. In practice outputs tend to be much longer than inputs. Inputs are human authored
		// whereas outputs are more likely to be produced by a machine (e.g. log output) and therefore very long
		maxInputLength = int(math.Floor(0.5*float64(maxLength)) + 1)
		maxOutputLength = maxInputLength
	}

	switch block.GetKind() {
	case v1alpha1.BlockKind_CODE:
		// Code just gets written as a code block
		sb.WriteString("```" + BASHLANG + "\n")

		data := block.GetContents()
		if len(data) > maxInputLength && maxInputLength > 0 {
			data = tailLines(data, maxInputLength)
			data = codeTruncationMessage + "\n" + data

			remaining := maxLength - len(data)
			if remaining > 0 {
				maxOutputLength += remaining
			}
		}
		sb.WriteString(data)
		sb.WriteString("\n```\n")
	default:
		// Otherwise assume its a markdown block

		data := block.GetContents()
		if len(data) > maxInputLength && maxInputLength > 0 {
			data = tailLines(data, maxInputLength)
			remaining := maxLength - len(data)
			if remaining > 0 {
				maxOutputLength += remaining
			}
		}
		sb.WriteString(data + "\n")
	}

	// Handle the outputs
	for _, output := range block.GetOutputs() {
		for _, oi := range output.Items {
			if oi.GetMime() == StatefulRunmeOutputItemsMimeType || oi.GetMime() == StatefulRunmeTerminalMimeType {
				// See: https://github.com/jlewi/foyle/issues/286. This output item contains a JSON dictionary
				// with a bunch of meta information that seems specific to Runme/stateful and not necessarily
				// relevant as context for AI so we filter it out. The output item we are interested in should
				// have a mime type of application/vnd.code.notebook.stdout and contain the stdout of the executed
				// code.
				//
				// We use an exclude list for now because Runme is adding additional mime types as it adds custom
				// renderers. https://github.com/stateful/vscode-runme/blob/3e36b16e3c41ad0fa38f0197f1713135e5edb27b/src/constants.ts#L6
				// So for now we want to error on including useless data rather than silently dropping useful data.
				// In the future we may want to revisit that.
				//
				continue
			}

			sb.WriteString("```" + OUTPUTLANG + "\n")
			textData := oi.GetTextData()
			if 0 < maxOutputLength && len(textData) > maxOutputLength {
				textData = textData[:maxOutputLength]
				sb.WriteString(textData)
				// Don't write a newline before writing truncation because that is more likely to lead to confusion
				// because people might not realize the line was truncated.
				// Emit a message indicating that the output was truncated
				// This is intended for the LLM so it knows that it is working with a truncated output.
				sb.WriteString(truncationMessage)
			} else {
				sb.WriteString(textData)
			}

			sb.WriteString("\n```\n")
		}
	}
}

// BlocksToMarkdown converts a sequence of blocks to markdown
func BlocksToMarkdown(blocks []*v1alpha1.Block) string {
	sb := strings.Builder{}

	for _, block := range blocks {
		writeBlockMarkdown(&sb, block, -1)
	}

	return sb.String()
}

// DocToMarkdown converts a doc to markdown
func DocToMarkdown(doc *v1alpha1.Doc) string {
	return BlocksToMarkdown(doc.GetBlocks())
}

// MarkdownToBlocks converts a markdown string into a sequence of blocks.
// This function relies on RunMe's Markdown->Cells conversion; underneath the hood that uses goldmark to walk the AST.
// RunMe's deserialization function doesn't have any notion of output in markdown. However, in Foyle outputs
// are rendered to code blocks of language "output". So we need to do some post processing to convert the outputs
// into output items
func MarkdownToBlocks(mdText string) ([]*v1alpha1.Block, error) {
	// N.B. We don't need to add any identities
	resolver := identity.NewResolver(identity.UnspecifiedLifecycleIdentity)
	notebook, err := editor.Deserialize([]byte(mdText), resolver)

	blocks := make([]*v1alpha1.Block, 0, len(notebook.Cells))

	var lastCodeBlock *v1alpha1.Block
	for _, cell := range notebook.Cells {

		var tr *parserv1.TextRange

		if cell.TextRange != nil {
			tr = &parserv1.TextRange{
				Start: uint32(cell.TextRange.Start),
				End:   uint32(cell.TextRange.End),
			}
		}

		cellPb := &parserv1.Cell{
			Kind:       parserv1.CellKind(cell.Kind),
			Value:      cell.Value,
			LanguageId: cell.LanguageID,
			Metadata:   cell.Metadata,
			TextRange:  tr,
		}

		block, err := converters.CellToBlock(cellPb)
		if err != nil {
			return nil, err
		}

		// We need to handle the case where the block is an output code block.
		if block.Kind == v1alpha1.BlockKind_CODE {
			if block.Language == OUTPUTLANG {
				// This is an output block
				// We need to append the output to the last code block
				if lastCodeBlock != nil {
					if lastCodeBlock.Outputs == nil {
						lastCodeBlock.Outputs = make([]*v1alpha1.BlockOutput, 0, 1)
					}
					lastCodeBlock.Outputs = append(lastCodeBlock.Outputs, &v1alpha1.BlockOutput{
						Items: []*v1alpha1.BlockOutputItem{
							{
								TextData: block.Contents,
							},
						},
					})
					continue
				}

				// Since we don't have a code block to add the output to just treat it as a code block
			} else {
				// Update the lastCodeBlock
				lastCodeBlock = block
			}
		} else {
			// If we have a non-nil markup block then we zero out lastCodeBlock so that a subsequent output block
			// wouldn't be added to the last code block.
			if block.GetContents() != "" {
				lastCodeBlock = nil
			}
		}

		blocks = append(blocks, block)
	}

	return blocks, err
}
